File: solutions/code-carl/ch09_dp/T4_LC0062_UniquePaths.java
/**
 * Practice template for:
 * - LeetCode 62. Unique Paths
 *   https://leetcode.cn/problems/unique-paths/
 */
import java.util.Arrays;

/**
 * @author Cheng Yihao
 * @version 1.0
 * @date 2024/11/20 16:32
 * @comment
 */
public class T4_LC0062_UniquePaths {
    public int uniquePaths(int m, int n) {
        int[] dp = new int[n];
        Arrays.fill(dp, 1);
        for (int i = 1; i < m; i++) {
            for (int j = 1; j < n; j++) {
                dp[j] += dp[j - 1];
            }
        }
        return dp[n - 1];
    }
}

File: solutions/python/code-carl/ch09_dp/t4_lc0062_unique_paths.py
"""LeetCode 62. Unique Paths
https://leetcode.cn/problems/unique-paths/
"""
from typing import List


class Solution:
    def uniquePaths(self, m: int, n: int) -> int:
        dp: List[List[int]] = [[0] * n for _ in range(m)]
        for i in range(m):
            dp[i][0] = 1
        for j in range(n):
            dp[0][j] = 1

        for i in range(1, m):
            for j in range(1, n):
                dp[i][j] = dp[i - 1][j] + dp[i][j - 1]

        return dp[-1][-1]


if __name__ == "__main__":
    solution = Solution()
    assert solution.uniquePaths(3, 7) == 28
    assert solution.uniquePaths(3, 2) == 3
    print("All tests passed.")
